sycl/doc/extensions/proposed/sycl_ext_oneapi_range_type.asciidoc

```diff
@@ -0,0 +1,168 @@
+
+This extension proposes a new kernel property that allows developers to declare
+the range requirements of individual kernels, providing more fine-grained
+control than existing compiler options and improved error behavior.
+
+The property described in this extension is an advanced feature that most
+applications should not need to use.
+In most cases, applications get the best performance without using this
+property.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_RANGE_TYPE` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New kernel property
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+struct range_type_key {
+  template <typename T>
+  using value_t = property_value<range_type_key, T>;
+};
+
+template <typename T>
+inline constexpr range_type_key::value_t<T> range_type;
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
+|===
+|Property|Description
+
+|`range_type`
+|The `range_type` property adds the requirement that the kernel must be
+compatible with kernel launches where the linear index of a work-item lies
+in the range [0, `std::numeric_limits<T>::max()`).
+
+If the implementation cannot satisfy this requirement, the implementation
+must throw an `exception` with the `errc::kernel_not_supported` error code,
+regardless of launch configuration.
+If the implementation can satisfy this requirement, but the kernel is
+launched with an incompatible configuration, the implementation must throw
+an `exception` with the `errc::nd_range` error code.
+
+`T` must be an integral type.
+
+|===
+
+== Usage example
+
+```c++
+namespace syclex = sycl::ext::oneapi::experimental;
+
+struct SmallKernel
+{
+  // Declare that this kernel supports [0, 2^31-1) work-items.
+  auto get(syclex::properties_tag) const {
+    return syclex::properties{syclex::range_type<int>};
+  }
+};
+
+struct LargeKernel
+{
+  // Declare that this kernel supports [0, 2^64-1) work-items.
+  auto get(syclex::properties_tag) const {
+    return syclex::properties{syclex::range_type<size_t>};
+  }
+};
+
+...
+
+// Throws an exception with errc::nd_range error code.
+// (because 2147483648 > 2147483647)
+q.parallel_for(2147483648, SmallKernel());
+
+// May throw an exception with errc::kernel_not_supported error code.
+// (if implementation/device doesn't support 64-bit ranges)
+q.parallel_for(2147483648, LargeKernel());
```

Comment:
I don't think we should allow this to fail, at least not for this reason.  The core spec says that the iteration variable size is `size_t`, so I think all conformant implementations must support this.  For nd-range kernels, there are various queries that the application needs to abide by regarding the nd-range.  It's possible that these queries disallow a global range that is 64-bits.  However, an implementation should not reject a kernel launch just because the iteration variable is `size_t`.

Reply:
> However, an implementation should not reject a kernel launch just because the iteration variable is `size_t`.

Maybe the issue here is really the name of this feature.  I don't see it as setting the type of the iteration variable (that will always be `size_t`, because that's what SYCL says) but rather declaring the range of work-item IDs that an implementation must support for this kernel.   I'm only using a type because typing something like `int` is easier than typing 2147483647.

Reply:
I think the change in c53b75bc747ea02d4d6f04290d400c0495ebb7ab renders this discussion obsolete.